File: minicode/manage_cli.py
from __future__ import annotations

from minicode.config import get_mcp_config_path, load_scoped_mcp_servers, save_scoped_mcp_servers
from minicode.skills import discover_skills, install_skill, remove_managed_skill


def _print_usage() -> None:
    print(
        "minicode management commands\n\n"
        "minicode mcp list [--project]\n"
        "minicode mcp add <name> [--project] [--protocol <auto|content-length|newline-json>] [--env KEY=VALUE ...] -- <command> [args...]\n"
        "minicode mcp remove <name> [--project]\n\n"
        "minicode skills list\n"
        "minicode skills add <path-to-skill-or-dir> [--name <name>] [--project]\n"
        "minicode skills remove <name> [--project]\n\n"
        "minicode valid-config"
    )


def _parse_scope(args: list[str]) -> tuple[str, list[str]]:
    rest = list(args)
    if "--project" in rest:
        rest.remove("--project")
        return "project", rest
    return "user", rest


def _take_option(args: list[str], name: str) -> str | None:
    if name not in args:
        return None
    index = args.index(name)
    if index + 1 >= len(args):
        raise RuntimeError(f"Missing value for {name}")
    value = args[index + 1]
    del args[index : index + 2]
    return value


def _take_repeat_option(args: list[str], name: str) -> list[str]:
    values: list[str] = []
    while name in args:
        index = args.index(name)
        if index + 1 >= len(args):
            raise RuntimeError(f"Missing value for {name}")
        values.append(args[index + 1])
        del args[index : index + 2]
    return values


def _parse_env_pairs(values: list[str]) -> dict[str, str]:
    env: dict[str, str] = {}
    for entry in values:
        if "=" not in entry:
            raise RuntimeError(f"Invalid --env value: {entry}")
        key, value = entry.split("=", 1)
        if not key.strip():
            raise RuntimeError(f"Invalid --env value: {entry}")
        env[key.strip()] = value
    return env


def _handle_mcp_command(cwd: str, args: list[str]) -> bool:
    if not args:
        _print_usage()
        return True
    subcommand, *rest_args = args
    scope, rest = _parse_scope(rest_args)
    if subcommand == "list":
        servers = load_scoped_mcp_servers(scope, cwd)
        if not servers:
            print(f"No MCP servers configured in {get_mcp_config_path(scope, cwd)}.")
            return True
        for name, server in servers.items():
            args_summary = " ".join(server.get("args", []))
            protocol = f" protocol={server['protocol']}" if server.get("protocol") else ""
            print(f"{name}: {server['command']} {args_summary}{protocol}".strip())
        return True
    if subcommand == "add":
        if "--" not in rest:
            raise RuntimeError("Use `--` before the MCP command.")
        separator_index = rest.index("--")
        head = rest[:separator_index]
        command_parts = rest[separator_index + 1 :]
        if not head or not command_parts:
            raise RuntimeError("Missing MCP server name or command.")
        name = head.pop(0)
        protocol = _take_option(head, "--protocol")
        env = _parse_env_pairs(_take_repeat_option(head, "--env"))
        if head:
            raise RuntimeError(f"Unknown arguments: {' '.join(head)}")
        command, *command_args = command_parts
        existing = load_scoped_mcp_servers(scope, cwd)
        existing[name] = {
            "command": command,
            "args": command_args,
            "env": env or None,
            "protocol": protocol,
        }
        save_scoped_mcp_servers(scope, existing, cwd)
        print(f"Added MCP server {name} to {get_mcp_config_path(scope, cwd)}")
        return True
    if subcommand == "remove":
        if not rest:
            raise RuntimeError("Missing MCP server name.")
        name = rest[0]
        existing = load_scoped_mcp_servers(scope, cwd)
        if name not in existing:
            print(f"MCP server {name} not found in {get_mcp_config_path(scope, cwd)}")
            return True
        del existing[name]
        save_scoped_mcp_servers(scope, existing, cwd)
        print(f"Removed MCP server {name} from {get_mcp_config_path(scope, cwd)}")
        return True
    _print_usage()
    return True


def _handle_skills_command(cwd: str, args: list[str]) -> bool:
    if not args:
        _print_usage()
        return True
    subcommand, *rest_args = args
    scope, rest = _parse_scope(rest_args)
    if subcommand == "list":
        skills = discover_skills(cwd)
        if not skills:
            print("No skills discovered.")
            return True
        for skill in skills:
            print(f"{skill.name}: {skill.description} ({skill.path})")
        return True
    if subcommand == "add":
        if not rest:
            raise RuntimeError("Missing skill source path.")
        source_path = rest[0]
        name = _take_option(rest, "--name")
        result = install_skill(cwd, source_path, name=name, scope=scope)
        print(f"Installed skill {result['name']} at {result['targetPath']}")
        return True
    if subcommand == "remove":
        if not rest:
            raise RuntimeError("Missing skill name.")
        result = remove_managed_skill(cwd, rest[0], scope=scope)
        if not result["removed"]:
            print(f"Skill {rest[0]} not found at {result['targetPath']}")
            return True
        print(f"Removed skill {rest[0]} from {result['targetPath']}")
        return True
    _print_usage()
    return True


def maybe_handle_management_command(cwd: str, argv: list[str]) -> bool:
    if not argv:
        return False
    category, *rest = argv
    if category == "mcp":
        return _handle_mcp_command(cwd, rest)
    if category == "skills":
        return _handle_skills_command(cwd, rest)
    if category in {"valid-config", "validate-config"}:
        from minicode.config import format_config_diagnostic
        print(format_config_diagnostic(cwd))
        return True
    if category in {"help", "--help", "-h"}:
        _print_usage()
        return True
    return False

File: ts-src/src/manage-cli.ts
import {
  type McpConfigScope,
  type McpServerConfig,
  MINI_CODE_MCP_TOKENS_PATH,
  getMcpConfigPath,
  loadScopedMcpServers,
  readMcpTokensFile,
  saveMcpTokensFile,
  saveScopedMcpServers,
} from './config.js'
import { discoverSkills, installSkill, removeManagedSkill } from './skills.js'

function printUsage(): void {
  console.log(`minicode management commands

minicode mcp list [--project]
minicode mcp add <name> [--project] [--protocol <auto|content-length|newline-json|streamable-http>] [--url <endpoint>] [--header KEY=VALUE ...] [--env KEY=VALUE ...] [-- <command> [args...]]
minicode mcp login <name> --token <bearer-token>
minicode mcp logout <name>
minicode mcp remove <name> [--project]

minicode skills list
minicode skills add <path-to-skill-or-dir> [--name <name>] [--project]
minicode skills remove <name> [--project]`)
}

function parseScope(args: string[]): {
  scope: McpConfigScope
  rest: string[]
} {
  const rest = [...args]
  const projectIndex = rest.indexOf('--project')
  if (projectIndex !== -1) {
    rest.splice(projectIndex, 1)
    return { scope: 'project', rest }
  }
  return { scope: 'user', rest }
}

function takeOption(args: string[], name: string): string | undefined {
  const index = args.indexOf(name)
  if (index === -1) return undefined
  const value = args[index + 1]
  if (!value) {
    throw new Error(`Missing value for ${name}`)
  }
  args.splice(index, 2)
  return value
}

function takeRepeatOption(args: string[], name: string): string[] {
  const values: string[] = []
  while (true) {
    const index = args.indexOf(name)
    if (index === -1) break
    const value = args[index + 1]
    if (!value) {
      throw new Error(`Missing value for ${name}`)
    }
    values.push(value)
    args.splice(index, 2)
  }
  return values
}

function parseEnvPairs(values: string[]): Record<string, string> {
  const env: Record<string, string> = {}
  for (const entry of values) {
    const separator = entry.indexOf('=')
    if (separator === -1) {
      throw new Error(`Invalid --env value: ${entry}`)
    }
    const key = entry.slice(0, separator).trim()
    const value = entry.slice(separator + 1)
    if (!key) {
      throw new Error(`Invalid --env value: ${entry}`)
    }
    env[key] = value
  }
  return env
}

async function handleMcpCommand(cwd: string, args: string[]): Promise<boolean> {
  const [subcommand, ...restArgs] = args
  if (!subcommand) {
    printUsage()
    return true
  }

  const { scope, rest } = parseScope(restArgs)

  if (subcommand === 'list') {
    const servers = await loadScopedMcpServers(scope, cwd)
    if (Object.keys(servers).length === 0) {
      console.log(`No MCP servers configured in ${getMcpConfigPath(scope, cwd)}.`)
      return true
    }

    for (const [name, server] of Object.entries(servers)) {
      const endpoint =
        server.url?.trim() ||
        `${server.command ?? ''} ${server.args?.join(' ') ?? ''}`.trim()
      const protocol = server.protocol ? ` protocol=${server.protocol}` : ''
      console.log(`${name}: ${endpoint}${protocol}`.trim())
    }
    return true
  }

  if (subcommand === 'add') {
    const separatorIndex = rest.indexOf('--')
    const head = separatorIndex === -1 ? [...rest] : rest.slice(0, separatorIndex)
    const commandParts = separatorIndex === -1 ? [] : rest.slice(separatorIndex + 1)
    const name = head.shift()
    if (!name) {
      throw new Error('Missing MCP server name.')
    }

    const protocol = takeOption(head, '--protocol') as McpServerConfig['protocol']
    const url = takeOption(head, '--url')?.trim()
    const env = parseEnvPairs(takeRepeatOption(head, '--env'))
    const headers = parseEnvPairs(takeRepeatOption(head, '--header'))
    if (head.length > 0) {
      throw new Error(`Unknown arguments: ${head.join(' ')}`)
    }

    const hasUrl = Boolean(url)
    const hasCommand = commandParts.length > 0
    if (hasUrl && hasCommand) {
      throw new Error('Cannot set both --url and local command. Choose one.')
    }
    if (!hasUrl && !hasCommand) {
      throw new Error('Missing MCP command or --url.')
    }
    if (protocol === 'streamable-http' && !hasUrl) {
      throw new Error('Protocol streamable-http requires --url.')
    }

    const [command = '', ...commandArgs] = commandParts
    const existing = await loadScopedMcpServers(scope, cwd)
    existing[name] = {
      command,
      args: hasCommand ? commandArgs : undefined,
      env: Object.keys(env).length > 0 ? env : undefined,
      url: hasUrl ? url : undefined,
      headers: Object.keys(headers).length > 0 ? headers : undefined,
      protocol,
    }
    await saveScopedMcpServers(scope, existing, cwd)
    console.log(`Added MCP server ${name} to ${getMcpConfigPath(scope, cwd)}`)
    return true
  }

  if (subcommand === 'remove') {
    const name = rest[0]
    if (!name) {
      throw new Error('Missing MCP server name.')
    }
    const existing = await loadScopedMcpServers(scope, cwd)
    if (!(name in existing)) {
      console.log(`MCP server ${name} not found in ${getMcpConfigPath(scope, cwd)}`)
      return true
    }
    delete existing[name]
    await saveScopedMcpServers(scope, existing, cwd)
    console.log(`Removed MCP server ${name} from ${getMcpConfigPath(scope, cwd)}`)
    return true
  }

  if (subcommand === 'login') {
    const name = rest[0]
    if (!name) {
      throw new Error('Missing MCP server name.')
    }
    const token = takeOption(rest, '--token')?.trim()
    if (!token) {
      throw new Error('Missing --token value.')
    }
    if (rest.length > 1) {
      throw new Error(`Unknown arguments: ${rest.slice(1).join(' ')}`)
    }
    const tokens = await readMcpTokensFile()
    tokens[name] = token
    await saveMcpTokensFile(tokens)
    console.log(`Stored MCP token for ${name} in ${MINI_CODE_MCP_TOKENS_PATH}`)
    return true
  }

  if (subcommand === 'logout') {
    const name = rest[0]
    if (!name) {
      throw new Error('Missing MCP server name.')
    }
    const tokens = await readMcpTokensFile()
    if (!(name in tokens)) {
      console.log(`No token found for ${name} in ${MINI_CODE_MCP_TOKENS_PATH}`)
      return true
    }
    delete tokens[name]
    await saveMcpTokensFile(tokens)
    console.log(`Removed MCP token for ${name} from ${MINI_CODE_MCP_TOKENS_PATH}`)
    return true
  }

  printUsage()
  return true
}

async function handleSkillsCommand(cwd: string, args: string[]): Promise<boolean> {
  const [subcommand, ...restArgs] = args
  if (!subcommand) {
    printUsage()
    return true
  }

  const { scope, rest } = parseScope(restArgs)

  if (subcommand === 'list') {
    const skills = await discoverSkills(cwd)
    if (skills.length === 0) {
      console.log('No skills discovered.')
      return true
    }
    for (const skill of skills) {
      console.log(`${skill.name}: ${skill.description} (${skill.path})`)
    }
    return true
  }

  if (subcommand === 'add') {
    const sourcePath = rest[0]
    if (!sourcePath) {
      throw new Error('Missing skill source path.')
    }
    const name = takeOption(rest, '--name')
    const result = await installSkill({
      cwd,
      sourcePath,
      name,
      scope,
    })
    console.log(`Installed skill ${result.name} at ${result.targetPath}`)
    return true
  }

  if (subcommand === 'remove') {
    const name = rest[0]
    if (!name) {
      throw new Error('Missing skill name.')
    }
    const result = await removeManagedSkill({
      cwd,
      name,
      scope,
    })
    if (!result.removed) {
      console.log(`Skill ${name} not found at ${result.targetPath}`)
      return true
    }
    console.log(`Removed skill ${name} from ${result.targetPath}`)
    return true
  }

  printUsage()
  return true
}

export async function maybeHandleManagementCommand(
  cwd: string,
  argv: string[],
): Promise<boolean> {
  const [category, ...rest] = argv
  if (!category) {
    return false
  }

  if (category === 'mcp') {
    return handleMcpCommand(cwd, rest)
  }

  if (category === 'skills') {
    return handleSkillsCommand(cwd, rest)
  }

  if (category === 'help' || category === '--help' || category === '-h') {
    printUsage()
    return true
  }

  return false
}
